### PR TITLE
Hotfix/anonymise columns

### PR DESCRIPTION
**Pull request by mereba**
This PR builds on @kieranajp's work on anonymsing columns. The current dump with anonymised fields (except the primary key, which we will later seed) looks like this:
```
# *******************************
# This database was nicked by Klepto™.
#
# https://github.com/hellofresh/klepto
# Host: 51f68e8bb931
# Database: fromDb
# Dumped at: Sat, 04 Nov 2017 06:18:29 +0100
# *******************************

SET NAMES utf8;
SET FOREIGN_KEY_CHECKS = 0;


CREATE TABLE `users` (
  `id` int(11) unsigned NOT NULL AUTO_INCREMENT,
  `username` varchar(255) DEFAULT '' COMMENT 'email address',
  `name` varchar(255) DEFAULT '' COMMENT 'first name',
  `password` varchar(255) DEFAULT '' COMMENT 'password',
  PRIMARY KEY (`id`)
) ENGINE=InnoDB AUTO_INCREMENT=8 DEFAULT CHARSET=utf8;


INSERT INTO `users` (`id`, `username`, `name`, `password`) VALUES
(%!s(*interface {}=0xc42007b540), "6West@Latz.edu", "Ashley", "1234"),
(%!s(*interface {}=0xc42007b9e0), "8Thompson@Kare.com", "Ryan", "1234"),
(%!s(*interface {}=0xc42007bd70), "iste_temporibus_explicabo@Jayo.gov", "Nicole", "1234"),
(%!s(*interface {}=0xc420240100), "yFuller@Youopia.org", "Samuel", "1234"),
(%!s(*interface {}=0xc420240460), "vHunter@Skilith.com", "Henry", "1234"),
(%!s(*interface {}=0xc4202407f0), "rTucker@Flashdog.org", "Cheryl", "1234"),
(%!s(*interface {}=0xc420240b50), "RalphLewis@Tagtune.biz", "Melissa", "1234");%
```

### CODE REVIEW

File: database/database.go [hunk 1]
```diff
@@ -4,6 +4,7 @@ import (
 	"database/sql"
 	"io"
 
+	// TODO: Why do we import mysql here?
```
Comment (kieranajp) [inline database/database.go]: It's required as a plugin for database/sql

File: database/mysql_anonymiser.go [hunk 2]
```diff
@@ -61,6 +54,37 @@ func (a *MySQLAnonymiser) DumpTable(table string, rowChan chan<- []*Cell, endCha
 	return nil
 }
 
+// - If the replacement is a literal, then we simply place it in the Cell. No faking here.
+// - If the replacement is not a literal, and it is an actual replacement then call the faker function
+// that it maps on to.
+// - We do not anonymise fields in columns that are not listed in config.
+func (a *MySQLAnonymiser) anonymiseCellv2(column, replacement string) *Cell {
```
Comment (kieranajp) [inline database/mysql_anonymiser.go]: just rename this and remove the other one

File: database/mysql_anonymiser.go [hunk 3]
```diff
@@ -61,6 +54,37 @@ func (a *MySQLAnonymiser) DumpTable(table string, rowChan chan<- []*Cell, endCha
...
+
+	if a.TypeScanner.Scan(literal); len(literal) != len(replacement) {
+		return &Cell{Column: column, Type: a.Detected, Value: literal}
+	}
+
+	if replacement != "" && len(literal) == len(replacement) {
+		value := (utils.Functions[replacement]).Call([]reflect.Value{})[0]
+		return &Cell{Column: column, Type: a.Detected, Value: value}
+	}
+	return nil
```
Comment (kieranajp) [inline database/mysql_anonymiser.go]: perhaps we should return an error instead

File: database/database.go (hunk 1 above)
Comment (mereba) [reply database/database.go]: Okay.

File: database/mysql_anonymiser.go (hunk 2 above)
Comment (mereba) [reply database/mysql_anonymiser.go]: Done.

File: database/mysql_anonymiser.go (hunk 3 above)
Comment (mereba) [reply database/mysql_anonymiser.go]: Sounds good, done.

**Comment by mereba** [comment]
We now have a version that preserves primary keys :)
```
# *******************************
# This database was nicked by Klepto™.
#
# https://github.com/hellofresh/klepto
# Host: 1d933542a06e
# Database: fromDb
# Dumped at: Mon, 06 Nov 2017 01:13:23 +0100
# *******************************

SET NAMES utf8;
SET FOREIGN_KEY_CHECKS = 0;


CREATE TABLE `users` (
  `id` int(11) unsigned NOT NULL AUTO_INCREMENT,
  `username` varchar(255) DEFAULT '' COMMENT 'email address',
  `name` varchar(255) DEFAULT '' COMMENT 'first name',
  `password` varchar(255) DEFAULT '' COMMENT 'password',
  PRIMARY KEY (`id`)
) ENGINE=InnoDB AUTO_INCREMENT=8 DEFAULT CHARSET=utf8;


INSERT INTO `users` (`id`, `username`, `name`, `password`) VALUES
(1, "HaroldSanders@Babbleset.mil", "Peter", "1234"),
(2, "incidunt@Edgeify.net", "William", "1234"),
(3, "AdamPeters@Pixonyx.org", "Keith", "1234"),
(4, "wHenderson@Edgetag.com", "Jeremy", "1234"),
(5, "natus@Flipbug.mil", "Donna", "1234"),
(6, "RogerTucker@Yakitri.biz", "Stephanie", "1234"),
(7, "wFuller@Midel.biz", "Diana", "1234");
```

File: database/mysql_anonymiser.go [hunk 4]
```diff
@@ -21,50 +24,94 @@ func NewMySQLAnonymiser(conn *sql.DB) *MySQLAnonymiser {
...
-func (a *MySQLAnonymiser) anonymiseCell(column, replacement string, scanner *utils.TypeScanner) *Cell {
-	if replacement != "" && scanner.Detected != "null" {
-		if literal := strings.TrimPrefix(replacement, literalPrefix); len(literal) != len(replacement) {
-			return &Cell{Column: column, Type: scanner.Detected, Value: literal}
-		}
-
-		return &Cell{Column: column, Type: scanner.Detected, Value: utils.Functions[replacement].Call(nil)[0]}
+// Useful for keeping primary key value unchanged. At this point, it is important for us
+// to not include in the config file the primary key column name.
+// TODO: warn user if primary key is listed as anonymised in config file.
```
Comment (kieranajp) [inline database/mysql_anonymiser.go]: Actually, we'll need a different config item in the file for the primary key. It won't always be called `id`; in fact in the bob DB it never is.

Comment (mereba) [reply database/mysql_anonymiser.go]: Sounds good. Removed the comment. The current approach is probably simpler. We preserve the values of any columns we have not explicitly specified to be anonymised. So a user will not even need extra config for the primary key. :)

File: database/mysql_anonymiser.go [hunk 5]
```diff
@@ -19,39 +24,46 @@ func NewMySQLAnonymiser(conn *sql.DB) *MySQLAnonymiser {
 	return &MySQLAnonymiser{conn: conn}
 }
 
-// DumpTable grabs the data from the provided database table and runs Faker against some columns
+// DumpTable grabs the data from the provided database table and runs Faker against
+// columns specified in config file.
 func (a *MySQLAnonymiser) DumpTable(table string, rowChan chan<- []*Cell, endChan chan<- bool) error {
 	rows, _ := a.conn.Query(fmt.Sprintf("SELECT * FROM `%s`", table))
```
Comment (rafaeljesus) [inline database/mysql_anonymiser.go]: Missing error handling

File: database/mysql_anonymiser.go [hunk 6]
```diff
@@ -19,39 +24,46 @@ func NewMySQLAnonymiser(conn *sql.DB) *MySQLAnonymiser {
...
 	defer rows.Close()
 
 	columns, _ := rows.Columns()
```
Comment (rafaeljesus) [inline database/mysql_anonymiser.go]: here too

File: database/mysql_anonymiser.go [hunk 7]
```diff
@@ -19,39 +24,46 @@ func NewMySQLAnonymiser(conn *sql.DB) *MySQLAnonymiser {
...
+				// Currently not very efficient because in each loop,
+				// we create len(columns) number of interfaces so that rows.Scan() will not complain
+				// about not having enough params. in reality, we only need
+				// len(columns) - numberOfAnonymisedFields interfaces
+				nFields := a.shouldNotAnonymise(columns)
+				if err := rows.Scan(nFields...); err != nil {
+					return err
+				}
+				seed := reflect.ValueOf(nFields[idx]).Elem()
+				cell, _ := a.keepsSeedValueUnchanged(column, seed, reflect.TypeOf(seed).Kind())
```
Comment (rafaeljesus) [inline database/mysql_anonymiser.go]: Missing handle error here too

File: database/mysql_anonymiser.go [hunk 8]
```diff
@@ -61,6 +73,43 @@ func (a *MySQLAnonymiser) DumpTable(table string, rowChan chan<- []*Cell, endCha
...
+	return nil, fmt.Errorf("couldn't anonymise cell with column: %v", column)
+}
+
 func (a *MySQLAnonymiser) shouldAnonymise(table, column string) string {
 	return viper.GetString(fmt.Sprintf("anonymise.%s.%s", table, column))
 }
+
+// Useful for keeping primary key or any other non-anonymous fields unchanged.
+func (a *MySQLAnonymiser) keepsSeedValueUnchanged(column string, value interface{}, typ interface{}) (*Cell, error) {
+	cell := &Cell{Column: column, Value: value, Type: fmt.Sprintf("%s", reflect.TypeOf(value).Kind())}
```
Comment (rafaeljesus) [inline database/mysql_anonymiser.go]: Perhaps it's more readable to have:
```go
kind := fmt.Sprintf("%s", reflect.TypeOf(value).Kind())
cell := &Cell{Column: column, Value: value, Type: kind}
```

File: database/mysql_anonymiser.go [hunk 9]
```diff
@@ -61,6 +73,43 @@ func (a *MySQLAnonymiser) DumpTable(table string, rowChan chan<- []*Cell, endCha
...
+	if cell.Type != "" {
+		return cell, nil
+	}
+	return nil, fmt.Errorf("couldn't keep cell value unchaged for column: %v", column)
+}
+
+func (a *MySQLAnonymiser) shouldNotAnonymise(columns []string) []interface{} {
+
```
Comment (rafaeljesus) [inline database/mysql_anonymiser.go]: Extra empty line

Comment (mereba) [reply database/mysql_anonymiser.go]: Removed it.

File: database/mysql_anonymiser.go (hunk 8 above)
Comment (mereba) [reply database/mysql_anonymiser.go]: Done.

File: database/mysql_anonymiser.go (hunk 7 above)
Comment (mereba) [reply database/mysql_anonymiser.go]: Done.

File: database/mysql_anonymiser.go (hunk 6 above)
Comment (mereba) [reply database/mysql_anonymiser.go]: Done

File: database/mysql_anonymiser.go (hunk 5 above)
Comment (mereba) [reply database/mysql_anonymiser.go]: Done

**Comment by rafaeljesus** [comment]
👍

**Comment by kieranajp** [comment]
:shipit:

**Pull request merged**